source/agora/flash/OnionPacket.d

```diff
@@ -101,11 +116,139 @@ public struct Payload
+    to obfuscate its true size. When a node peels of their layer of the
+    encrypted packet, it adds additional pading to fill the packet size
+    back to its expected fixed length size.
+
+    Params:
+        payment_hash = the payment hash to use
+        lock_height = the initial lock height
+        amount = the amount for the payment
+        path = the individual hops (including destination hop)
+        total_amount = will contain the amount which needs to be paid to the
+            first channel along the route. Different to `amount` as it also
+            includes fees.
+
+    Returns:
+        the onion packet ready to be routed through the first payment path.
+
+*******************************************************************************/
+
+public OnionPacket createOnionPacket (in Hash payment_hash,
+    in Height lock_height, in Amount amount, in Hop[] path,
+    out Amount total_amount)
+{
+    assert(path.length >= 1);
+
+    // todo: setting fees should be part of the routing algorithm
+    total_amount = amount;
+    foreach (hop; path)
+    {
+        if (!total_amount.add(hop.fee))
+            assert(0);
+    }
+
+    Amount forward_amount = total_amount;
+    Height outgoing_lock_height = lock_height;
+    Hash next_chan_id;
+
+    Pair ephemeral_kp = Pair.random();
+    OnionPacket packet = { version_byte : 0 };
+
+    // onion packets have to be created from right to left
+    assert(path.length <= 20);
+    ulong last_index = path.length - 1;
+
+    foreach (hop; path.retro)
+    {
+        Payload payload =
+        {
+            next_chan_id : next_chan_id,
+            forward_amount : forward_amount,
+            outgoing_lock_height : outgoing_lock_height,
+        };
+
+        auto encrypted_payload = encryptPayload(payload, ephemeral_kp,
+            hop.pub_key);
+
+        packet.encrypted_payloads[last_index] = encrypted_payload;
+        last_index--;
+        if (!forward_amount.sub(hop.fee))
+            assert(0);
+
+        // todo: use htlc_delta config here from the channel config
+        assert(outgoing_lock_height != 0);
+        outgoing_lock_height = Height(outgoing_lock_height - 1);
+
+        next_chan_id = hop.chan_id;
+
+        // keep updating last valid ephemeral key
+        packet.ephemeral_pk = ephemeral_kp.V;
+
+        // todo: use hashing for derivation instead
+        auto next_secret = ephemeral_kp.v + Scalar(hashFull(1));
+        ephemeral_kp = Pair(next_secret, next_secret.toPoint());
+    }
+
+    // fill out the rest with encrypted filler
+    foreach (ref payload; packet.encrypted_payloads[path.length .. $])
+        fillGarbage(payload);
+
+    return packet;
+}
+
+/// Fill the payload with random encrypted data so it looks real but it ain't
+private void fillGarbage (ref EncryptedPayload payload)
+{
+    auto ephemeral_kp = Pair.random();
+
+    ubyte[32] key_bytes = ephemeral_kp.v[][0 .. 32];
+    randombytes_buf_deterministic(payload.nonce.ptr,
```

Comment:
This should be fixed. The randomness of this may not generate a nonce that's on the curve. So it's better to just use `Scalar.random()` here.

Reply:
Ah no I was wrong, this is not a point on the curve. But it is just a random value, so we can just use `randombytes_buf` for this.

Reply:
Fixed.